src/main/java/org/terasology/metalrenegades/interaction/systems/WellWaterSystem.java

```diff
@@ -15,36 +15,146 @@
+        EntityRef heldItem = gatheringCharacter.getComponent(CharacterHeldItemComponent.class).selectedItem;
+        if (!gatheringCharacter.exists()) {
+            return;
+        }
+
+        EntityRef wellEntity = getWellEntity(sourceBlock);
+        if (wellEntity == null) {
+            logger.warn("No well entity found for activated well block!");
+            return;
+        }
+
+        WellSourceComponent wellSourceComp = wellEntity.getComponent(WellSourceComponent.class);
+        if (wellSourceComp.waterRefills <= 0) { // if no refills remain, don't give water.
+            return;
+        }
+        wellSourceComp.waterRefills--;
+        wellEntity.saveComponent(wellSourceComp);
+
+        if (!heldItem.hasComponent(WaterCupComponent.class)) {
+            ThirstComponent thirst = gatheringCharacter.getComponent(ThirstComponent.class);
+
+            thirst.lastCalculatedWater = thirst.maxWaterCapacity;
+            thirst.lastCalculationTime = time.getGameTimeInMs();
+            gatheringCharacter.saveComponent(thirst);
+
+            wellEntity.send(new WellDrinkEvent(gatheringCharacter));
 
-        if(!cupItem.exists() || !gatheringCharacter.exists()) {
             return;
         }
 
+        EntityRef cupItem = entityManager.create("MetalRenegades:waterCup");
         cupItem.send(new GiveItemEvent(gatheringCharacter));
+        wellEntity.send(new CupFilledEvent(gatheringCharacter, cupItem));
+        heldItem.destroy();
+    }
+
+    @ReceiveEvent(components = {WaterCupComponent.class})
+    public void onDrinkConsumed(DrinkConsumedEvent event, EntityRef item) {
+        EntityRef drinkingCharacter = event.getInstigator();
+        EntityRef cupItem = entityManager.create("MetalRenegades:emptyCup");
+
+        if (!cupItem.exists() || !drinkingCharacter.exists()) {
+            return;
+        }
+
+        item.destroy();
+        cupItem.send(new GiveItemEvent(drinkingCharacter));
+    }
+
+    /**
+     * Retrieves the well entity associated with a well water source block.
+     *
+     * @param sourceBlock The activated source block.
+     * @return The well entity that contains this block.
+     */
+    private EntityRef getWellEntity(EntityRef sourceBlock) {
+        LocationComponent blockLocComp = sourceBlock.getComponent(LocationComponent.class);
+        Vector3i blockLocation = new Vector3i(blockLocComp.getWorldPosition());
+        EntityRef wellEntity = null;
+
+        for (EntityRef waterSource : entityManager.getEntitiesWith(WellSourceComponent.class)) {
+            DynParcelRefComponent dynParcelRefComponent = waterSource.getComponent(DynParcelRefComponent.class);
+            Rect2i parcelRect = dynParcelRefComponent.dynParcel.getShape();
+
+            if (parcelRect.contains(blockLocation.x, blockLocation.z)) {
+                wellEntity = waterSource;
+                break;
+            }
+        }
+
+        return wellEntity;
```

Comment:
With a more functional approach to using Stream API this could be written as

```suggestion
        
        return StreamSupport.stream(entityManager.getEntitiesWith(WellSourceComponent.class).spliterator(), false)
                .filter(waterSource -> containsLocation(waterSource, blockLocation))
                .findFirst()
                .orElse(EntityRef.NULL);
```

The method `containsLocation` probably needs a better name, but I don't fully get what is happening there to be honest 🙈 

```java
    private boolean containsLocation(EntityRef waterSource, Vector3i location) {
        DynParcelRefComponent dynParcelRefComponent = waterSource.getComponent(DynParcelRefComponent.class);
        Rect2i parcelRect = dynParcelRefComponent.dynParcel.getShape();
        return parcelRect.contains(location.x, location.z);
    }
```

Reply:
... or with using a `Predicate<EntityRef>`:

```suggestion

        return StreamSupport.stream(entityManager.getEntitiesWith(WellSourceComponent.class).spliterator(), false)
                .filter(atLocation(blockLocation))
                .findFirst()
                .orElse(EntityRef.NULL);
```

and

```java
    private Predicate<EntityRef> atLocation(Vector3i location) {
        return (waterSource) -> {
            DynParcelRefComponent dynParcelRefComponent = waterSource.getComponent(DynParcelRefComponent.class);
            Rect2i parcelRect = dynParcelRefComponent.dynParcel.getShape();
            return parcelRect.contains(location.x, location.z);
        };
    }
```

Reply:
Woah, the stream implementation looks really nice! I'm still not very well versed in using streams over ordinary loops, I'll need to do some research on that. 👍